pkg/apply/basic_printer.go

```diff
@@ -49,82 +50,139 @@ func (a *applyStats) sum() int {
-	pruneCount := 0
-	deleteCount := 0
+	statusCollector := &statusCollector{
+		latestStatus: make(map[object.ObjMetadata]pollevent.Event),
+		printStatus:  false,
+	}
+	pruneStats := &pruneStats{}
+	deleteStats := &deleteStats{}
 	for e := range ch {
 		switch e.Type {
 		case event.ErrorType:
 			cmdutil.CheckErr(e.ErrorEvent.Err)
 		case event.ApplyType:
-			ae := e.ApplyEvent
-			if ae.Type == event.ApplyEventCompleted {
-				output := fmt.Sprintf("%d resource(s) applied. %d created, %d unchanged, %d configured",
-					applyStats.sum(), applyStats.created, applyStats.unchanged, applyStats.configured)
-				// Only print information about serverside apply if some of the
-				// resources actually were applied serverside.
-				if applyStats.serversideApplied > 0 {
-					output += fmt.Sprintf(", %d serverside applied", applyStats.serversideApplied)
-				}
-				printFunc(output)
-			} else {
-				obj := ae.Object
-				gvk := obj.GetObjectKind().GroupVersionKind()
-				name := getName(obj)
-				applyStats.inc(ae.Operation)
-				printFunc("%s %s", resourceIDToString(gvk.GroupKind(), name),
-					strings.ToLower(ae.Operation.String()))
-			}
+			b.processApplyEvent(e.ApplyEvent, applyStats, statusCollector, printFunc)
 		case event.StatusType:
-			statusEvent := e.StatusEvent
-			switch statusEvent.EventType {
-			case pollevent.ResourceUpdateEvent:
-				id := statusEvent.Resource.Identifier
-				gk := id.GroupKind
-				printFunc("%s is %s: %s", resourceIDToString(gk, id.Name),
-					statusEvent.Resource.Status.String(), statusEvent.Resource.Message)
-			case pollevent.ErrorEvent:
-				id := statusEvent.Resource.Identifier
-				gk := id.GroupKind
-				printFunc("%s error: %s\n", resourceIDToString(gk, id.Name),
-					statusEvent.Error.Error())
-			case pollevent.CompletedEvent:
-				printFunc("all resources has reached the Current status")
-			case pollevent.AbortedEvent:
-				printFunc("resources failed to the reached Current status")
-			}
+			b.processStatusEvent(e.StatusEvent, statusCollector, printFunc)
 		case event.PruneType:
-			pe := e.PruneEvent
-			if pe.Type == event.PruneEventCompleted {
-				printFunc("%d resource(s) pruned", pruneCount)
-			} else {
-				obj := e.PruneEvent.Object
-				gvk := obj.GetObjectKind().GroupVersionKind()
-				name := getName(obj)
-				pruneCount++
-				printFunc("%s %s", resourceIDToString(gvk.GroupKind(), name), "pruned")
-			}
+			b.processPruneEvent(e.PruneEvent, pruneStats, printFunc)
 		case event.DeleteType:
-			de := e.DeleteEvent
-			if de.Type == event.DeleteEventCompleted {
-				printFunc("%d resource(s) deleted", deleteCount)
-			} else {
-				obj := de.Object
-				gvk := obj.GetObjectKind().GroupVersionKind()
-				name := getName(obj)
-				deleteCount++
-				printFunc("%s %s", resourceIDToString(gvk.GroupKind(), name), "deleted")
-			}
+			b.processDeleteEvent(e.DeleteEvent, deleteStats, printFunc)
 		}
 	}
 }
 
+func (b *BasicPrinter) processApplyEvent(ae event.ApplyEvent, as *applyStats,
```

Comment:
It should be easier to write unit tests for each of these `processEvent` functions. We can do that in future PR's.